File: scripts/github-fy.js
#!/usr/bin/env node
/**
 * Obsidian → GitHub Markdown 转换器
 * 用法: node scripts/github-fy.js [--dry-run]
 */

const fs = require("fs");
const path = require("path");

const VAULT_ROOT = path.resolve(__dirname, "..");
const DOCS_DIR = path.join(VAULT_ROOT, "docs");

const NOTE_FILES = [
  "Javase.md",
  "数据结构.md",
  "排序算法详解.md",
  "agent知识点.md",
  "项目实战.md",
];

// ====== Frontmatter → collapsible details ======
function convertFrontmatter(content) {
  if (!content.startsWith("---")) return content;

  const end = content.indexOf("---", 3);
  if (end === -1) return content;

  const yamlStr = content.slice(3, end).trim();
  const body = content.slice(end + 3).replace(/^\n+/, "");

  if (!yamlStr) return body;

  // 简单 YAML 解析（支持单行 key:val 和多行数组）
  const lines = yamlStr.split("\n");
  const rows = [];
  let inArray = false, arrayKey = "", arrayVals = [];
  for (const line of lines) {
    // 多行数组项: "  - value"
    const arrMatch = line.match(/^\s+-\s+(.+)/);
    if (arrMatch && inArray) {
      arrayVals.push(arrMatch[1].trim());
      continue;
    }
    // 上一组数组结束，写入
    if (inArray && arrayVals.length > 0) {
      rows.push(`| ${arrayKey} | \`${arrayVals.join(", ")}\` |`);
      arrayVals = [];
      inArray = false;
    }
    // 新的 key: value 行
    const m = line.match(/^(\w[\w-]*):\s*(.*)/);
    if (m) {
      const key = m[1];
      let val = m[2].trim();
      if (val === "") {
        // 可能是多行数组开头
        inArray = true;
        arrayKey = key;
        continue;
      }
      // 去掉 YAML 方括号
      val = val.replace(/^\[|\]$/g, "");
      rows.push(`| ${key} | \`${val}\` |`);
    }
  }
  // 最后一组数组
  if (inArray && arrayVals.length > 0) {
    rows.push(`| ${arrayKey} | \`${arrayVals.join(", ")}\` |`);
  }

  if (rows.length === 0) return body;

  const details =
    "<details>\n" +
    "<summary>📋 元数据</summary>\n\n" +
    "| 属性 | 值 |\n" +
    "|------|----|\n" +
    rows.join("\n") +
    "\n\n</details>\n\n";

  return details + body;
}

// ====== Wikilink → Markdown link ======
function convertWikilinks(content) {
  // 保护代码块
  const blocks = [];
  content = content.replace(/```[\s\S]*?```/g, (m) => {
    blocks.push(m);
    return `%%BLOCK_${blocks.length - 1}%%`;
  });

  // 保护内联代码
  const inlines = [];
  content = content.replace(/`[^`]+`/g, (m) => {
    inlines.push(m);
    return `%%INLINE_${inlines.length - 1}%%`;
  });

  // ![[image.png]] → ![image.png](image.png)
  content = content.replace(
    /!\[\[([^\]]+)\]\]/g,
    (_, file) => `![${file}](${file})`
  );

  // [[note#heading|alias]]
  content = content.replace(
    /\[\[([^\]|#]+)#([^\]|]+)\|([^\]]+)\]\]/g,
    (_, note, heading, alias) =>
      `[${alias}](${note}.md#${slugify(heading)})`
  );

  // [[note#heading]]
  content = content.replace(
    /\[\[([^\]|#]+)#([^\]]+)\]\]/g,
    (_, note, heading) =>
      `[${note} - ${heading}](${note}.md#${slugify(heading)})`
  );

  // [[note|alias]]
  content = content.replace(
    /\[\[([^\]|]+)\|([^\]]+)\]\]/g,
    (_, note, alias) => `[${alias}](${note}.md)`
  );

  // [[note]]
  content = content.replace(
    /\[\[([^\]]+)\]\]/g,
    (_, note) => `[${note}](${note}.md)`
  );

  // 还原
  blocks.forEach((b, i) => {
    content = content.replace(`%%BLOCK_${i}%%`, b);
  });
  inlines.forEach((c, i) => {
    content = content.replace(`%%INLINE_${i}%%`, c);
  });

  return content;
}

function slugify(text) {
  return text
    .trim()
    .toLowerCase()
    .replace(/\s+/g, "-")
    .replace(/[^\w一-鿿-]/g, "");
}

// ====== 折叠式 callout → 普通 callout ======
function convertCallouts(content) {
  return content.replace(/> \[!(\w+)\]-/g, "> [!$1]");
}

// ====== 主流程 ======
function convertFile(filename, dryRun) {
  const src = path.join(VAULT_ROOT, filename);
  const dst = path.join(DOCS_DIR, filename);

  if (!fs.existsSync(src)) {
    console.log(`  ⚠ ${filename} — 文件不存在，跳过`);
    return false;
  }

  let content = fs.readFileSync(src, "utf-8");
  const original = content;

  content = convertFrontmatter(content);
  content = convertWikilinks(content);
  content = convertCallouts(content);

  if (content === original) {
    console.log(`  · ${filename} — 无需转换`);
  } else {
    if (!dryRun) {
      fs.mkdirSync(path.dirname(dst), { recursive: true });
      fs.writeFileSync(dst, content, "utf-8");
    }
    console.log(`  ✓ ${filename} → docs/${filename}`);
  }

  return true;
}

function main() {
  const dryRun = process.argv.includes("--dry-run");

  console.log("🔧 Obsidian → GitHub 转换器");
  console.log(`   源目录: ${VAULT_ROOT}`);
  console.log(`   输出目录: ${DOCS_DIR}`);
  if (dryRun) console.log("   [DRY RUN — 不写入文件]");
  console.log();

  if (!dryRun) fs.mkdirSync(DOCS_DIR, { recursive: true });

  for (const file of NOTE_FILES) {
    convertFile(file, dryRun);
  }

  console.log();
  console.log("✅ 完成！");
}

main();

File: scripts/github-fy.py
#!/usr/bin/env python3
"""
Obsidian → GitHub Markdown 转换器

处理：
- [[wikilink]]          → [wikilink](wikilink.md)
- [[wikilink|alias]]    → [alias](wikilink.md)
- [[wikilink#heading]]  → [wikilink - heading](wikilink.md#heading)
- ![[image.png]]        → ![image.png](image.png)
- ---frontmatter---     → <details><summary>📋 元数据</summary>...表格...</details>

用法：
  python3 scripts/github-fy.py              # 转换所有笔记到 docs/
  python3 scripts/github-fy.py --dry-run    # 预览不写入
"""

import re
import os
import sys
from pathlib import Path
import yaml  # pip install pyyaml

VAULT_ROOT = Path(__file__).parent.parent
DOCS_DIR = VAULT_ROOT / "docs"
NOTE_FILES = [
    "Javase.md",
    "数据结构.md",
    "排序算法详解.md",
    "agent知识点.md",
    "项目实战.md",
]


def convert_frontmatter(content: str) -> str:
    """将 YAML frontmatter 转为可折叠的 HTML details"""
    if not content.startswith("---"):
        return content

    end = content.find("---", 3)
    if end == -1:
        return content

    yaml_str = content[3:end].strip()
    body = content[end + 3:].lstrip()

    if not yaml_str:
        return body

    try:
        data = yaml.safe_load(yaml_str)
    except yaml.Error:
        return body

    if not isinstance(data, dict) or not data:
        return body

    # 构建表格
    rows = []
    for k, v in data.items():
        if isinstance(v, list):
            v = ", ".join(str(x) for x in v)
        elif v is None:
            v = ""
        rows.append(f"| {k} | {v} |")

    if not rows:
        return body

    details = (
        "<details>\n"
        "<summary>📋 元数据</summary>\n\n"
        "| 属性 | 值 |\n"
        "|------|------|\n"
        + "\n".join(rows) +
        "\n\n</details>\n\n"
    )

    return details + body


def convert_wikilinks(content: str) -> str:
    """转换 [[wikilink]] 为 GitHub 兼容的 Markdown 链接"""
    # 不转换代码块内的 wikilinks
    # 策略：先保护代码块，转换外部内容，再还原

    # 保护内联代码和代码块
    inline_codes = []
    code_blocks = []

    def save_inline(m):
        inline_codes.append(m.group(0))
        return f"%%INLINE_{len(inline_codes) - 1}%%"

    def save_block(m):
        code_blocks.append(m.group(0))
        return f"%%BLOCK_{len(code_blocks) - 1}%%"

    content = re.sub(r"`[^`]+`", save_inline, content)
    content = re.sub(r"```[\s\S]*?```", save_block, content)

    # 转换 ![[image.png]] → ![image.png](image.png)
    content = re.sub(
        r"!\[\[([^\]]+)\]\]",
        lambda m: f"![{m.group(1)}]({m.group(1)})",
        content,
    )

    # 转换 [[note#heading|alias]] → [alias](note.md#heading)
    content = re.sub(
        r"\[\[([^\]|#]+)#([^\]|]+)\|([^\]]+)\]\]",
        lambda m: f"[{m.group(3)}]({m.group(1)}.md#{_slugify(m.group(2))})",
        content,
    )

    # 转换 [[note#heading]] → [note - heading](note.md#heading)
    content = re.sub(
        r"\[\[([^\]|#]+)#([^\]]+)\]\]",
        lambda m: f"[{m.group(1)} - {m.group(2)}]({m.group(1)}.md#{_slugify(m.group(2))})",
        content,
    )

    # 转换 [[note|alias]] → [alias](note.md)
    content = re.sub(
        r"\[\[([^\]|]+)\|([^\]]+)\]\]",
        r"[\2](\1.md)",
        content,
    )

    # 转换 [[note]] → [note](note.md)
    content = re.sub(
        r"\[\[([^\]]+)\]\]",
        r"[\1](\1.md)",
        content,
    )

    # 还原代码块和内联代码
    for i, block in enumerate(code_blocks):
        content = content.replace(f"%%BLOCK_{i}%%", block)
    for i, code in enumerate(inline_codes):
        content = content.replace(f"%%INLINE_{i}%%", code)

    return content


def _slugify(text: str) -> str:
    """把中文标题转成 GitHub 锚点格式"""
    # GitHub 锚点：保留中文（直接作为 ID），其他非字母数字变连字符
    # 简化处理：去掉多余空格
    text = text.strip().lower()
    text = re.sub(r"\s+", "-", text)
    # 去掉非字母数字非中文非连字符的字符
    text = re.sub(r"[^\w一-鿿-]", "", text)
    return text


def convert_obsidian_callouts(content: str) -> str:
    """
    GitHub 已支持 Obsidian callouts (tip, note, warning, important)。
    但折叠式 callouts (>[!note]-) 不支持，转为普通 callout。
    """
    content = re.sub(r"> \[!(\w+)\]-", r"> [!\1]", content)
    return content


def convert_file(src: Path, dst: Path) -> bool:
    """转换单个文件"""
    if not src.exists():
        print(f"  ⚠ {src.name} — 文件不存在，跳过")
        return False

    content = src.read_text(encoding="utf-8")

    original = content
    content = convert_frontmatter(content)
    content = convert_wikilinks(content)
    content = convert_obsidian_callouts(content)

    if content == original:
        print(f"  ✓ {src.name} — 无需转换")
    else:
        dst.parent.mkdir(parents=True, exist_ok=True)
        dst.write_text(content, encoding="utf-8")
        print(f"  ✓ {src.name} → docs/{dst.name}")

    return True


def main():
    dry_run = "--dry-run" in sys.argv

    print("🔧 Obsidian → GitHub 转换器")
    print(f"   源目录: {VAULT_ROOT}")
    print(f"   输出目录: {DOCS_DIR}")
    if dry_run:
        print("   [DRY RUN — 不写入文件]")
    print()

    DOCS_DIR.mkdir(parents=True, exist_ok=True)

    for filename in NOTE_FILES:
        src = VAULT_ROOT / filename
        dst = DOCS_DIR / filename
        convert_file(src, dst)

    print()
    print("✅ 完成！docs/ 目录已生成 GitHub 友好版本。")
    print("   在 README.md 中链接到 docs/ 下的文件即可。")


if __name__ == "__main__":
    main()
